{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"machine_shape":"hm"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"},"gpuClass":"standard"},"cells":[{"cell_type":"markdown","source":["# Imports\n"],"metadata":{"id":"n-AQfSfrRsIf"}},{"cell_type":"code","execution_count":null,"metadata":{"id":"jQUpvmpz1BZZ"},"outputs":[],"source":["from google.colab import drive\n","drive.mount('/content/gdrive')"]},{"cell_type":"code","source":["import spacy\n","import json\n","!pip install rouge_score\n","from rouge_score import rouge_scorer\n","!pip install nltk\n","from nltk.translate.bleu_score import sentence_bleu\n","!pip install scientific_information_change\n","from scientific_information_change.estimate_similarity import SimilarityEstimator\n","!pip install sentence_transformers\n","from sentence_transformers import SentenceTransformer, util\n","import re"],"metadata":{"id":"blHUy4DT1UB9"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# Generate contexts"],"metadata":{"id":"M79Q_QPkR15x"}},{"cell_type":"code","source":["### Code block adapted from https://github.com/allenai/s2orc\n","# TODO: Loop through batch IDs to process all S2ORC data\n","BATCH_ID = 0\n","# create a lookup for the pdf parse based on paper ID\n","paper_id_to_pdf_parse = {}\n","with open(f'/content/gdrive/MyDrive/Independent study - Max & Carlos/data_sample_from_s2orc/pdf_parses_{BATCH_ID}.jsonl') as f_pdf:\n","    for line in f_pdf:\n","        pdf_parse_dict = json.loads(line)\n","        paper_id_to_pdf_parse[pdf_parse_dict['paper_id']] = pdf_parse_dict\n","print(\"PDFs parsed.\")\n","\n","# filter papers using metadata values\n","citation_contexts = []\n","failed_keys = {}\n","with open(f'/content/gdrive/MyDrive/Independent study - Max & Carlos/data_sample_from_s2orc/metadata_{BATCH_ID}.jsonl') as f_meta:\n","    for line in f_meta:\n","        metadata_dict = json.loads(line)\n","        paper_id = metadata_dict['paper_id']\n","        # Can add other filters here if needed\n","        if not metadata_dict['has_outbound_citations']:\n","            continue\n","        # get citation context (paragraphs)!\n","        if paper_id in paper_id_to_pdf_parse:\n","            # (1) get the full pdf parse from the previously computed lookup dict\n","            pdf_parse = paper_id_to_pdf_parse[paper_id]\n","            # (2) pull out fields we need from the pdf parse, including bibliography & text\n","            bib_entries = pdf_parse['bib_entries']\n","            # If citing paper has no data, move on\n","            if not pdf_parse['abstract'] or not pdf_parse['body_text']:\n","                continue\n","            paragraphs = pdf_parse['abstract'] + pdf_parse['body_text']\n","            # (3) loop over paragraphs, grabbing citation contexts\n","            for paragraph in paragraphs:\n","                # (4) loop over each inline citation in this paragraph\n","                for cite_span in paragraph['cite_spans']:\n","                    # (5) each inline citation can be resolved to a bib entry\n","                    try:\n","                        cited_bib_entry = bib_entries[cite_span['ref_id']]\n","                    except:\n","                        if paper_id in failed_keys:\n","                            failed_keys.get(paper_id).append(cite_span['ref_id'])\n","                        else:\n","                            failed_keys[paper_id] = [cite_span['ref_id']]\n","                    # (6) that bib entry *may* be linked to a S2ORC paper. if so, grab paragraph\n","                    linked_paper_id = cited_bib_entry['link']\n","                    if linked_paper_id and linked_paper_id in paper_id_to_pdf_parse and \\\n","                    paper_id_to_pdf_parse[linked_paper_id]['abstract'] and \\\n","                    paper_id_to_pdf_parse[linked_paper_id]['body_text'] and not \\\n","                    linked_paper_id == paper_id:\n","                        citation_contexts.append({\n","                            'citing_paper_id': paper_id,\n","                            'cited_paper_id': linked_paper_id,\n","                            'context': paragraph['text'],\n","                            'citation_mention_start': cite_span['start'],\n","                            'citation_mention_end': cite_span['end'],\n","                        })\n","print(\"Citation contexts created\")"],"metadata":{"id":"WFiNFBpl193v","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1675661365523,"user_tz":480,"elapsed":217551,"user":{"displayName":"Maxwell Bennett","userId":"04059206930111574135"}},"outputId":"76f79b73-43e8-4922-c721-e79b32e35732"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["PDFs parsed.\n","Citation contexts created\n"]}]},{"cell_type":"code","source":["# # To create a text file in G drive with given content\n","# from pydrive.drive import GoogleDrive\n","# from pydrive.auth import GoogleAuth\n","# gauth = GoogleAuth()\n","# # Create local webserver and auto handles authentication.\n","# gauth.LocalWebserverAuth()\n","# # Create GoogleDrive instance with authenticated GoogleAuth instance.\n","# drive = GoogleDrive(gauth)\n","# # Create GoogleDriveFile instance\n","# file = drive.CreateFile({'title':'failed_keys.txt', 'mimeType':'text/plain'})\n","content = \"\"\n","for key in failed_keys:\n","    content += key + \": \"\n","    for ref in failed_keys[key]:\n","        content += ref[6:] + \" \"\n","    content += \"\\n\"\n","print(content)\n","# file.SetContentString(content)\n","# file.Upload() # Upload file."],"metadata":{"id":"22NKQEwcj68G"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Load ROUGE scorer\n","scorer = rouge_scorer.RougeScorer(['rouge1', 'rouge2', 'rougeL'], use_stemmer=True)\n","# Load IMS\n","estimator = SimilarityEstimator()\n","# Load model for language processing\n","!pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.5.1/en_core_sci_sm-0.5.1.tar.gz\n","nlp = spacy.load('en_core_sci_sm')\n","# Load SBERT model\n","model = SentenceTransformer('nli-distilroberta-base-v2')\n","\n","#### We may need to add a check in that the full pdf parse of the cited paper is in the pdf_parse\n","#### Paragraph may need the ['text'] indexing\n","#### we should limit this for just a few papers for testing\n","\n","#### citation context = secondary paper\n","#### window size = primary paper\n","\n","### Hyperparameters\n","window_size = 2\n","context_type_paragraph = False\n","context_size = 2\n","context_offset = -1 # negative indicates shifted left. If context_size = 2 and offset = -1, then the citing context will be the sentence preceding the citing sentence and the citing sentence.\n","scoring_type = 'sbert'\n","threshold = False\n","k = 4\n"],"metadata":{"id":"ntLFFaVfLxXA"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Score the similarity of two strings given a scoring method\n","def similarity(scoring_type, str1, str2):\n","    scoring_type = scoring_type.lower()\n","    if (scoring_type == 'rouge'):  # 0-1\n","        scores = scorer.score(str1, str2)\n","        avg_score = (scores['rouge1'][2] + scores['rouge2'][2] + scores['rougeL'][2]) / 3\n","        return avg_score\n","    elif (scoring_type == 'bleu'):  # 0-1\n","        return sentence_bleu(str1.split(), str2.split(), weights=(0.25, 0.25, 0.25, 0.25))\n","    elif (scoring_type == 'ims' or scoring_type == 'information matching score'):  # 0-5, normalized to 0-1\n","        return estimator.estimate_ims_array([str1], [str2]) / 5\n","    elif (scoring_type == 'sbert'):  # 0-1\n","        embeddings1 = model.encode([str1], convert_to_tensor=True)\n","        embeddings2 = model.encode([str2], convert_to_tensor=True)\n","        return util.cos_sim(embeddings1, embeddings2).item()\n","    else:\n","        raise Exception(\"Please provide valid scoring_type\")\n","\n","# Sort windows and similarity scores, returning only scores/windows of interest (top-k or threshold)\n","def sort_windows(scores, windows, threshold, k):\n","    windows = [x for _, x in sorted(zip(scores, windows), key=lambda pair: pair[0], reverse=True)]\n","    scores.sort(reverse=True)\n","    if threshold:\n","        scores = [x for x in scores if x >= k]\n","        windows = windows[:len(scores)]\n","    else: # top-k\n","        scores = scores[:k]\n","        windows = windows[:k]\n","    return scores, windows"],"metadata":{"id":"Wn5yDIgEL2Hb"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Batch 0: ~11k valid, down from 3.5 million. This will improve when other batches are processed."],"metadata":{"id":"Mbca_cUUa_MY"}},{"cell_type":"code","source":["### TODO: Find distortion scores for all citations\n","\n","for i in range(20):\n","    cc = citation_contexts[i]\n","\n","# for cc in citation_contexts:\n","\n","    ### (1) Find citation's context_window:\n","    # Find the citing paper key (secondary)\n","    citing_id = cc['citing_paper_id']\n","    # Extract citation context (context size is hyperparam)\n","    if context_type_paragraph: # If context window is paragraph\n","        context_window = cc['context']\n","    else: # Context paragraph is something else\n","        # Collect PDF text of the citing paper\n","        pdf_parse = paper_id_to_pdf_parse[citing_id]\n","        paragraphs = pdf_parse['abstract'] + pdf_parse['body_text']\n","        citing_sentences = []\n","        # Loop through the entire text\n","        context_sent_index = 0\n","        for paragraph in paragraphs: # This can (and maybe should...) be done in preprocessing block\n","            doc = nlp(paragraph['text'])\n","            if paragraph['text'] == cc['context']:\n","                citation_start_index = cc['citation_mention_start']\n","                for sent in doc.sents:\n","                    if citation_start_index > sent.start_char and citation_start_index < sent.end_char:\n","                        context_sent_index = len(citing_sentences)\n","                    citing_sentences.append(sent.text)\n","            else:\n","                for sent in doc.sents:\n","                    citing_sentences.append(sent.text)\n","        try:\n","            context_window = ' '.join(citing_sentences[context_sent_index + context_offset : context_sent_index + context_offset + context_size]) # this may causes errors with offset? start of array. also end\n","        except:\n","            context_window = ' '.join(citing_sentences[context_sent_index : context_sent_index + context_size])\n","\n","    ### (2) Generate array of all sentences in the cited paper\n","    cited_id = cc['cited_paper_id']\n","    pdf_parse = paper_id_to_pdf_parse[cited_id]\n","    paragraphs = pdf_parse['abstract'] + pdf_parse['body_text']\n","    sentences = []\n","    for paragraph in paragraphs:\n","        doc = nlp(paragraph['text'])\n","        for sent in doc.sents:\n","            sentences.append(sent.text)\n","\n","    ### (3) Score all windows in cited/primary\n","    cited_windows = []\n","    cited_scores = []\n","    # Remove citation notations from context (sentence [2] example)\n","    context_window = re.sub(r' \\[\\d+\\]', '', context_window)\n","    # For each possible window in cited/primary:\n","    for i in range(len(sentences) - window_size + 1):\n","        # Extract window from cited paper and remove citation notation\n","        curr_window = ' '.join(sentences[i:i + window_size])\n","        curr_window = re.sub(r' \\[\\d+\\]', '', curr_window)\n","        # Conduct similarity assessment\n","        score = similarity(scoring_type, context_window, curr_window)\n","        cited_windows.append(curr_window)\n","        cited_scores.append(score)\n","\n","    ### (4) Determine windows in top-k/threshold\n","    cited_scores, cited_windows = sort_windows(cited_scores, cited_windows, threshold, k)\n","    print(\"\\nCitation context: \" + context_window)\n","    for i in range(len(cited_scores)):\n","        print(str(cited_scores[i]) + \":\\t\" + cited_windows[i])\n","\n","    ### (5) TODO: Find distortion scores for each window\n"],"metadata":{"id":"U5KJsE3a2_Ib","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1675581439687,"user_tz":480,"elapsed":934092,"user":{"displayName":"Maxwell Bennett","userId":"04059206930111574135"}},"outputId":"fb616b09-d602-444c-97d7-a0b44638907e"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["\n","Citation context: Furthermore, several mathematical models used to describe biological processes require a noise term to adequately model the behaviour of these processes. Cellular processes, such as transcription and translation, chromatin remodeling and pathwayspecific regulation, are sources of stochastic events leading to cell-to-cell variability , .\n","0.6957463026046753:\tHowever, the oscillatory properties, such as phase and amplitude are dependent on the gene function and differ between tissues and experimental conditions . Oscillatory patterns of expression in major housekeeping genes responsible for the energy balance (PPAR) and basic transcription (TBP) are bound to impose the same patterns on all transcribed genes regardless of the volume of transcription.\n","0.6850059032440186:\tSecond, analysis of correlation with phase shift (also used to identify phase groups) confirms high correlation of nearly all profiles to common cosine curves. Third, living cells are known to have more than one oscillator, but these oscillators are normally synchronized to the rhythm of the circadian molecular clock, active in peripheral tissues.\n","0.6778979301452637:\tThird, living cells are known to have more than one oscillator, but these oscillators are normally synchronized to the rhythm of the circadian molecular clock, active in peripheral tissues. Testing individual expression profiles for periodicity we are looking for manifestation of the same factor, hence not independent hypothesis.\n","0.6561210751533508:\tBackground: Microarrays are widely used for estimation of expression of thousands of genes in a biological sample. The resolution ability of this method is limited by the background noise.\n","\n","Citation context: By using the network mapping of edges to a physical system of 2N bosons, it is evident that eventually only half of them can \"reach\" the fundamental energy level of the network, which corresponds to the highest fitness present. The Kaniadakis κ-distribution , was originally introduced in 2002 and later re-obtained within the framework of a Jaynes Maximum Entropy principle , which starts from a generalized system entropy S κ given by\n","0.5990433692932129:\tIndeed 1/κ is the reciprocal of light speed while 1/κ 2 is the particle rest energy, in dimensionless forms. On the other hand the self-duality relation given by Eq. (7.4) leads to the relativistic dispersion relation, obtaining in this way a direct link with the microscopic dynamics of the system.\n","0.5928324460983276:\tWe indicate with W i the microscopic energy, with β the reciprocal of the temperature and with µ the chemical potential, so that, the Lagrange multipliers are given by a 0 = βµ and a 1 = β. The system entropy is given by 10) while the maximum entropy principle yields the probability distribution function\n","0.5644968748092651:\tThe deformation mechanism introduced by the parameter κ is induced directly by the Lorentz transformations of Einstein's special relativity. Indeed 1/κ is the reciprocal of light speed while 1/κ 2 is the particle rest energy, in dimensionless forms.\n","0.553260326385498:\tWe note that the parameters λ and η produce a scaling to the Lagrange multipliers. Ordinary statistical mechanics corresponds to the choice Λ(x) = ln(x), E(x) = exp(x) and {α = 1, λ = 1, η = 1}, or alternatively {α = 1/e, λ = 1, η = 0}. We pose now the question regarding the possible existence of further couples of functions Λ(x), E(x) and set of parameters {α, λ, η} defining statistical theories, through Eqs. (1.1), (2.5) and (3.2), different from ordinary statistical mechanics.\n","\n","Citation context: with p l being the probability of the system being in state l and κ ∈ (−1, 1). It can be shown that the generalized logarithm ln κ (x) is self-scaling and self-dual , as the ordinary logarithm ln (x) of the Boltzmann-Shannon entropy\n","0.844066858291626:\ti) Self-dual logarithm: the choice corresponding to τ = 0, ς = 1 and −1 < κ < 1, yields the generalized logarithm ln κ (x) = ln 0 0 1 (x), which together with its inverse function assume the form The self-duality of the above functions is expressed through the relationships\n","0.8352599143981934:\tis sufficient to fix the form of Λ(x) and therefore the form of the entropy and of the probability distribution function. The obtained class of generalized logarithms, Λ(x) = ln κτ ς (x), depends on three free parameters according to\n","0.8302947282791138:\tThe ordinary logarithm, which is self-dual and self-scaling, can be obtained as the limiting case when κ → 0 and τ = 0, independently on the value of the parameter ς, such that we can write ln(x) = ln 0 0 ς (x). i) Self-dual logarithm: the choice corresponding to τ = 0, ς = 1 and −1 < κ < 1, yields the generalized logarithm ln κ (x) = ln 0 0 1 (x), which together with its inverse function assume the form\n","0.8242678642272949:\tThe following integral representation for ln κ (x) holds ln κ (x) = 1 2 which in the limit of κ → 0 reduces to a well known property of the ordinary logarithm.\n","\n","Citation context: The parameter κ can be seen as a deviation from extensivity, as ln κ (x) recovers the extensive case when κ → 0. The inverse of the generalized logarithm is the generalized exponential [11, 17] ,\n","0.8362821340560913:\tiv) Only in very few cases, the generalized logarithm given by Eq. (3.5), can be inverted to obtain analytically the corresponding generalized exponential. For instance the choice κ = 3γ/2, τ = γ/2, ς = 1, yields an invertible generalized logarithm with y = 1 − 4γ 3 x 3 , .\n","0.8296483755111694:\tOf course, the generalized exponential function defining the probability distribution function after inversion, produces a generalized logarithm Λ(pi). But, in general, the mean value of −Λ(pi) is not the entropy of the system.\n","0.8085997104644775:\tAfter noting that when −|κ| ≤ τ ≤ |κ| it holds d ln κτ ς (x)/dx > 0, we can define the three-parameter exponential exp κτ ς (x) as the inverse function of ln κτ ς (x). The properties of exp κτ ς (x) readily follow from those of ln κτ ς (x).\n","0.8082196712493896:\twhith τ ′ = −τ and ς ′ = 1/ς. After noting that when −|κ| ≤ τ ≤ |κ| it holds d ln κτ ς (x)/dx > 0, we can define the three-parameter exponential exp κτ ς (x) as the inverse function of ln κτ ς (x).\n","\n","Citation context: We can solve the above differential equation by applying a mean field approximation, in which we replace the partition sum Z κ by its average Z κ , For this purpose we introduce a chemical potential µ and a mean fugacity z κ = e βµ κ and hence by using the thermodynamical stability of the deformed κ-exponential one obtains:\n","0.8181209564208984:\tWe indicate with W i the microscopic energy, with β the reciprocal of the temperature and with µ the chemical potential, so that, the Lagrange multipliers are given by a 0 = βµ and a 1 = β. The system entropy is given by 10) while the maximum entropy principle yields the probability distribution function\n","0.8109575510025024:\tIn order to better explain how the functions exp κ (x) and ln κ (x) can define a statistical theory we consider briefly the paradigm of statistical mechanics. We indicate with W i the microscopic energy, with β the reciprocal of the temperature and with µ the chemical potential, so that, the Lagrange multipliers are given by a 0 = βµ and a 1 = β.\n","0.8012984991073608:\tFrom the characteristic equations we obtain the system 1 ± τ + κ = λ α −τ ∓κ , which can be solved to determine the two constants α and λ. The constant η follows from Eqs. The conditions Λ(0 + ) < 0 and 0 + Λ(0 + ) = 0 imply that τ ≤ |κ| and τ > |κ|−1 respectively.\n","0.7857130765914917:\tThe latter equation is a first order differentialfunctional equation whose solutions determine unambiguously the form of the generalized logarithm after taking into account the conditions (i) Λ(1) = 0, (ii) 0 + Λ(0 + ) = 0 + and (iii) Λ ′ (1) = 1. After obtaining the generalized logarithm and by inversion the generalized exponential, through Eqs.(1.1) and (3.2), the entropy and the probability distribution function are unambiguously fixed.\n","\n","Citation context: these results recover what is already known in the extensive limit . By the definition of z k and exploiting the self-duality of the κ-exponential , we have that\n","0.7272133827209473:\tfor any value of x. After taking into account the definition of the density entropy, we get the following property for the generalized logarithm The latter relationship can be verified easily by starting with Eq. (3.4) and after taking into account that for −|κ| ≤ τ ≤ |κ| it results d ln κτ ς (x)/dx > 0.\n","0.7016042470932007:\tThe latter inequality and Eq. (5.7) imply for any value of x. After taking into account the definition of the density entropy, we get the following property for the generalized logarithm\n","0.6894969940185547:\tWe consider here the two-parameter generalized logarithms ln κς (x) = ln κ 0 ς (x), obtained by posing τ = 0 in Eq. (3.5) . It is easy to verify that ln κς (x) is the scaled function of ln κ (x).\n","0.6889770030975342:\twhich in the limit of κ → 0 reduces to a well known property of the ordinary logarithm. The properties of the functions ln κ (x) and exp κ (x) have been study extensively in the literature due to their relevance in relativistic statistical mechanics.\n","\n","Citation context: which, in our context, represents a deformed Bose-Einstein distribution, with the very same properties of monotonicity as the original Bose-Einstein distribution n ( ) in the non-extensive case. Contrary to the case of the fitness model generalized with the Tsallis entropy , the Bose-Einstein distribution n κ ( ) can be rewritten in a more insightful way by using properties of the κ-deformed exponential [11, 17] .\n","0.708446741104126:\tIt is clear that the form of Λ(x) imposes the form of the distribution function according to the maximum entropy principle. In proposing generalized statistical theories some standard properties (positivity, continuity, concavity, symmetry, etc.) are customarily required for the entropy.\n","0.6945524215698242:\tThe function Λ(x) can be viewed as a generalization of the ordinary logarithmic function. We remark that the entropy is the ordinary mean value of −Λ(p i ).\n","0.6925846338272095:\tregardless on the form of the generalized entropy. In the following, we are interested to the particular class of probability distribution functions which are given in terms of the generalized exponential function\n","0.6900851130485535:\tRegarding the meaning of Eq. (3.4) we note that the equation guarantees a twofold link between entropy and the probability distribution function: firstly we have a differential link offered by the maximum entropy principle through Eq. (2.5); secondly we have an algebraic link offered by Eq. (1.1) and Eq. (3.2) where the same function in direct and inverse form, i.e. Λ(x) and E(x), defines the entropy and the probability distribution function according to the standard rules of ordinary statistical mechanics. Eq. (3.4) has been proposed in ref. , while the special case corresponding to η = 0 has been considered in ref. .\n","\n","Citation context: These findings demonstrate that human FMs respond to two viral signatures by generating distinct inflammatory cytokine/chemokine profiles and antiviral responses through different mechanisms. Bacterial infection and inflammation of the placental fetal membranes (FM) play a major role in chorioamnionitis, preterm premature rupture of membranes (PPROM) and preterm birth (ACOG Practice Bulletin No. 80, 2007 , Goldenberg et al., 2000 Lamont, 2003; Caughey et al., 2008; Goldenberg et al., 2008) .\n","0.7858489155769348:\t66, 67 The mechanisms by which intrauterine infections lead to preterm labour are related to activation of the innate immune system. 68 Microorganisms are recognised by pattern-recognition receptors-eg, toll-like receptors, which in turn elicit the release of infl ammatory chemokines and cytokines-such as interleukin 8, interleukin 1β, and tumour necrosis factor (TNF) α.\n","0.7850611805915833:\tProstaglandins stimulate uterine contractility, whereas degradation of extracellular matrix in the fetal membranes leads to PPROM. 67, 68 Microbiological studies suggest that intrauterine infection might account for 25-40% of preterm births; 67 however, 25-40% might be a minimum estimate because intrauterine infection is diffi cult to detect with conventional culture techniques.\n","0.7839369773864746:\t89 Both studies suggest that subclinical fetal infection is far more common than traditionally recognised. Microbial invasion of the amniotic cavity is frequently associated with intra-amniotic infl ammation and a fetal infl ammatory response.\n","0.7656591534614563:\tMicrobial invasion of the amniotic cavity is frequently associated with intra-amniotic infl ammation and a fetal infl ammatory response. 90, 91 The fetal infl ammatory response has been linked to the onset of preterm labour, and fetal injury and long-term handicap-including periventricular leucomalacia, cerebral palsy, and chronic lung disease.\n","\n","Citation context: In contrast, neither Poly(I:C) nor ssRNA induced a significant increase in SLPI mRNA levels (Fig. 4) . Infection-associated pregnancy complications such as chorioamnionitis, PPROM and preterm birth have been strongly associated with bacterial infection and inflammation of the FMs (ACOG Practice Bulletin No. 80, 2007 , Goldenberg et al., 2000 Lamont, 2003; Caughey et al., 2008; Goldenberg et al., 2008) .\n","0.7643171548843384:\tProstaglandins stimulate uterine contractility, whereas degradation of extracellular matrix in the fetal membranes leads to PPROM. 67, 68 Microbiological studies suggest that intrauterine infection might account for 25-40% of preterm births; 67 however, 25-40% might be a minimum estimate because intrauterine infection is diffi cult to detect with conventional culture techniques.\n","0.7378370761871338:\t113, 114 In several studies, the viral DNAs-identifi ed by PCR techniques-in the amniotic fl uid of asymptomatic women undergoing genetic amniocentesis were generally unrelated to subsequent preterm births; 115 therefore, it seems unlikely that maternal viral infection plays an important part in preterm birth, but controversy persists, and with little information, further study is needed. 116 Several studies have shown an association between uterine contraction frequency and preterm birth; however, uterine contractions do not predict preterm birth well in singletons because of the wide variation in frequency in normal pregnancy and the large overlap in frequency between women who do and do not deliver preterm.\n","0.7368718385696411:\tCompared with bacterial infections, there is sparse evidence that viral infections predispose to preterm birth; however, when the mother is severely ill, such as with varicella pneumonia or severe acute respiratory syndrome, a preterm delivery might occur. 113, 114 In several studies, the viral DNAs-identifi ed by PCR techniques-in the amniotic fl uid of asymptomatic women undergoing genetic amniocentesis were generally unrelated to subsequent preterm births; 115 therefore, it seems unlikely that maternal viral infection plays an important part in preterm birth, but controversy persists, and with little information, further study is needed.\n","0.7357513904571533:\t78, 79 At 21-24 weeks' gestation, most spontaneous births are associated with histological chorioamnionitis compared with about 10% at 35-36 weeks. 78, 79 The microorganisms most commonly reported in the amniotic cavity are genital Mycoplasma spp, and, specifi cally, U urealyticum, but many other organisms have been identifi ed. Some common lower genital tract microorganisms, such as Streptococcus agalactiae, are rarely seen in the amniotic cavity before membrane rupture.\n","\n","Citation context: Until recently, the limitations of nucleic acid-encoded library technologies were largely due to necessary in vivo steps, which limit library size and provide selection pressure that cannot be controlled by the experimenter. A lately introduced method for the preparation of a minimal genetic unit which consists of a mRNA covalently linked to a polypeptide (PROfusion TM ) circumvents the in vivo step and allows the selection for peptides and proteins with desired properties totally in vitro .\n","0.7732720375061035:\tIn general, such approaches are limited by the complexity of the sequence libraries that can be generated; for example, phage display libraries typically are limited by transfection efficiency to less than 10 9 independent members. More recently, selection schemes based on the display of the nascent peptide chain on the surface of the ribosome have been developed (15) (16) (17) .\n","0.7622055411338806:\t9 and 10), plasmid display (11) , and completely in vivo genetic approaches (12) (13) (14) . In general, such approaches are limited by the complexity of the sequence libraries that can be generated; for example, phage display libraries typically are limited by transfection efficiency to less than 10 9 independent members.\n","0.743438720703125:\tThe main barrier to the development of effective methods for protein evolution has been the difficulty of recovering the information encoding a protein sequence after the protein has been translated. Until recently, most approaches to this problem have involved a step in which the DNA is transcribed and translated in vivo, and the resulting protein is expressed in such a way as to remain physically linked to the encoding nucleic acid, which then is recovered for amplification and further selection.\n","0.7419826984405518:\tThis approach has the advantage of being fully in vitro and potentially allowing larger libraries (Ͼ10 12 ) to be explored; however, selections must be performed under conditions that preserve the integrity of the ribosome⅐mRNA⅐peptide ternary complex. We sought to develop a simpler and more robust system in which an mRNA would become directly attached to the peptide or protein it encodes by a stable covalent linkage.\n","\n","Citation context: RNAprotein fusion libraries with up to 10 14 different sequences have been generated and have been successfully used for the isolation of peptide and protein binders to a variety of different targets . The originally described method for RNA-protein fusion preparation involved translation of an enzymatically linked mRNA-puromycin conjugate .\n","0.8416925668716431:\tCovalent RNA-peptide fusions should provide an additional route to the in vitro selection and directed evolution of proteins. In vitro selection experiments using RNA and DNA have shown that nucleic acid molecules with specific molecular recognition and catalytic properties can be isolated from complex pools of random sequences by repeated rounds of selection and amplification (for reviews see refs. .\n","0.8252411484718323:\tA higher fraction of the input RNA was converted to fusion product in the presence of a higher ribosome͞template ratio; over the range of input mRNA concentrations that we tested approximately 0.8-1.0 ϫ 10 12 fusion molecules are made per ml of translation extract. For RNA-peptide fusions to be useful, the peptide attached to an mRNA must be the peptide encoded by that mRNA, i.e., the nascent peptide must not be transferred to the puromycin of some other mRNA.\n","0.8108950853347778:\tThe stable linkage between the informational (nucleic acid) and functional (peptide) domains of the resulting joint molecules allows a specific mRNA to be enriched from a complex mixture of mRNAs based on the properties of its encoded peptide. Fusions between a synthetic mRNA and its encoded myc epitope peptide have been enriched from a pool of random sequence mRNA-peptide fusions by immunoprecipitation.\n","0.8094037771224976:\tCurrent protocols are consistent with the generation of mRNA-peptide fusion libraries consisting of 10 12 -10 13 independent members. We have generated a library of 10 12 fusions, in which the peptide domain contains a random 27-aa sequence, in a 10-ml translation reaction.\n","\n","Citation context: We found the above protocol to be useful for efficient synthesis of mRNA-peptide and mRNAprotein fusion libraries from a number of template libraries different in length and complexity. Efficient mRNA affinity purification on neutravidin agarose carrying the biotinylated puromycin-linker replaced the laborious mRNA gel purification step used in previous protocols [3, 5] .\n","0.7877011299133301:\tThis approach has the advantage of being fully in vitro and potentially allowing larger libraries (Ͼ10 12 ) to be explored; however, selections must be performed under conditions that preserve the integrity of the ribosome⅐mRNA⅐peptide ternary complex. We sought to develop a simpler and more robust system in which an mRNA would become directly attached to the peptide or protein it encodes by a stable covalent linkage.\n","0.7842819690704346:\tA higher fraction of the input RNA was converted to fusion product in the presence of a higher ribosome͞template ratio; over the range of input mRNA concentrations that we tested approximately 0.8-1.0 ϫ 10 12 fusion molecules are made per ml of translation extract. For RNA-peptide fusions to be useful, the peptide attached to an mRNA must be the peptide encoded by that mRNA, i.e., the nascent peptide must not be transferred to the puromycin of some other mRNA.\n","0.7691713571548462:\tSynthesis of mRNA-Peptide Fusions. We began with the synthesis of an appropriately protected form of puromycin that could be coupled to a CPG solid support.\n","0.7642096281051636:\tCovalent RNA-peptide fusions should provide an additional route to the in vitro selection and directed evolution of proteins. In vitro selection experiments using RNA and DNA have shown that nucleic acid molecules with specific molecular recognition and catalytic properties can be isolated from complex pools of random sequences by repeated rounds of selection and amplification (for reviews see refs. .\n","\n","Citation context: The induction of aB-crystallin, an anti-apoptotic and anti-oxidant protein [4, , by genotoxic stress may also serve a homeostatic function to suppress apoptosis and reduce ROS levels in the setting of DNA damage. Indeed, recent studies indicate that p53 induces the expression of several pro-survival and anti-oxidant genes, including COX-2, the sestrins, TIGAR, glutathione peroxidase 1, aldehyde dehydrogenase 4 family member A1, and receptor tyrosine kinase DDR1 .\n","0.8730276823043823:\tModulation of sestrin expression by p53 might be part of an adaptive response mechanism that protects the cell against oxidative stress. In addition, substantial variations in the expression of PA26 and Hi95 in different tissues (18, 20) might be responsible for tissue-specific differences in peroxide-dependent signaling and for variability in the effects of elevated ROS and RNS.\n","0.8516781330108643:\tWe show that sestrins, a family of proteins whose expression is modulated by p53, are required for regeneration of Prxs containing Cys-SO 2 H, thus reestablishing the antioxidant firewall. Sestrins contain a predicted redox-active domain homologous to AhpD, the enzyme catalyzing the reduction of a bacterial Prx, AhpC. Purified Hi95 (sestrin 2) protein supports adenosine triphosphate-dependent reduction of overoxidized PrxI in vitro, indicating that unlike AhpD, which is a disulfide reductase, sestrins are cysteine sulfinyl reductases.\n","0.8516781330108643:\tWe show that sestrins, a family of proteins whose expression is modulated by p53, are required for regeneration of Prxs containing Cys-SO 2 H, thus reestablishing the antioxidant firewall. Sestrins contain a predicted redox-active domain homologous to AhpD, the enzyme catalyzing the reduction of a bacterial Prx, AhpC. Purified Hi95 (sestrin 2) protein supports adenosine triphosphate-dependent reduction of overoxidized PrxI in vitro, indicating that unlike AhpD, which is a disulfide reductase, sestrins are cysteine sulfinyl reductases.\n","0.8405951261520386:\tWe have recently shown that overexpression of the p53-regulated gene Hi95 (18) , a member of the sestrin gene family (19) , protects cells against apoptosis induced by ischemia or H 2 O 2 , suggesting that Hi95 is involved in defense against ROS (18) . To gain insight into the functions of sestrins [Hi95, PA26 (20) , and Sesn3], we undertook a detailed sequence analysis of these proteins (21) .\n","\n","Citation context: All populations responded fairly equally to SDF-1 ␣ , confirming their competence to migrate under our experimental conditions (Fig. 1 C) . Interestingly, cells of pre-pro-B cell phenotype (B220 lo HSA Ϫ /lo AA4.1 ϩ NK1.1 Ϫ ; a subpopulation within the original Fraction A population of Hardy et al. that combines the A1 and A2 fractions ) displayed substantial migration to RANTES, JE (mouse MCP-1), MIG, and TECK, chemokines associated with monocyte and/or activated T cell responses in the periphery.\n","0.7305580377578735:\tWe found that the B220+S7+ cell fractions in BM are the only B-lineage populations capable of proliferation (and differentiation) on the stromal cell layer in short term (4-d) culture. As shown in Fig. 3 a, the 30171 + fractions (Fr. B, C) exhibited very significant proliferation during 4-d stromal culture compared with the 30171 -fraction (Fr.\n","0.7045173645019531:\tThe use of the recently described mAb S7, together with 13220 (RA3-6132) and multiparameter FACS® analysis clearly defines a population (3-4%) in BM highly enriched for the earliest B lineage cells. Thus, our analysis shows that B-lineage cells express S7 at a very early (pre-pro-B and pro-B cell) stage, then downregulate it upon VDJ rearrangement and transition to the pre-B cell stage.\n","0.7033680081367493:\tThe three distinct phenotypes of cells in this B220+S7+ cell population, 30171 -BP-1', 30171+BP-1 -, 30171+BP-1+ are termed Fr. A, B, and C (Fig. 2, left) .\n","0.685822606086731:\t11, 7 Activity to S7+ B Lineage Cells. We found that the B220+S7+ cell fractions in BM are the only B-lineage populations capable of proliferation (and differentiation) on the stromal cell layer in short term (4-d) culture.\n","\n","Citation context: In contrast, splenic ␦ ϩ follicular B cells were uniformly and strongly CXCR5 ϩ (Fig. 4 B, (Fig. 5 B) . Moreover, the responses of putative blood-borne peripheral B cells passing through the BM at the time of harvest (Fraction F in the classification scheme of Hardy et al. , B220 hi CD43 Ϫ sIgM ϩ ) were indistinguishable from those of splenic follicular ␦ ϩ B cells (data not shown).\n","0.7976137399673462:\tWe reasoned that the earliest B-committed cells could be found by investigating the correlated expression of determinants consistently present on B-lineage cells (such as B220; reference 5) together with those not found on surface IgM+ B cells, but possibly expressed on B progenitor stages . Simultaneously, we have determined levels of several other cell surface molecules whose expression is known to change during differentiation and which might facilitate resolution of heterogeneity within the B lineage cells in BM.\n","0.7848122119903564:\tSorted cells of each phenotype from BM were analyzed by FACS® for DNA content by propidium iodide staining. This analysis revealed that the most mature S7+B220+ fraction (Fr. C) contains significant numbers (>30%) of cells in cycle in adult BM (Fig. 5, left) and that such cycling cells comprise an increasing percentage in Fr.\n","0.7816779613494873:\tPre-B and B cells were defined as in Fig. 2 , right panel; that is, pre-B are S7 -B220+IgM -(Fr. D in the figure) and B cells are S7 -B220+IgM* (Fr.\n","0.7781612873077393:\tConsequently, these data allow us to demonstrate phenotypic and physiological alterations in early B lineage cells according to their differentiation stage (pre-pro-B, early pro-B, late pro-B), progressing from absolute dependence on cell contact mediated signals to depen- Murine Bone Marrow Pro-B Cells dence only on a factor present in the stromal supernatant, identifiable as IL7 .\n","\n","Citation context: B cell genesis derives from dividing hematopoietic stem cells (HSCs) that differentiate into progenitor cells, which in turn give rise to various hematopoietic lineages. The first identifiable BM cells committed to the B lineage are Fraction A1/A2 (AA4.1 ϩ NK1.1 Ϫ cells within Fraction A [25, 26, 40] ); these cells still have their Ig genes in the germline state but display the B lineage marker B220.\n","0.8529717326164246:\tSubsequently, such cells further differentiate to express BM and this coincides with alteration of the physiological characteristics of the B-lineage cells. Previous work with stromal cell cultures led Dorshkind to suggest that the earliest B lineage progenitors are absolutely dependent on signals mediated by direct contact with the layer while latter stages are relatively contact independent (42).\n","0.8251491189002991:\tConsequently, these data allow us to demonstrate phenotypic and physiological alterations in early B lineage cells according to their differentiation stage (pre-pro-B, early pro-B, late pro-B), progressing from absolute dependence on cell contact mediated signals to depen- Murine Bone Marrow Pro-B Cells dence only on a factor present in the stromal supernatant, identifiable as IL7 .\n","0.8149995803833008:\tOur finding that it is first expressed on a transitional stage of pro-B cells demonstrates its highly ordered regulation during B cell differentiation . Recently BP-1 has been cloned and sequenced and shown to be the second member 1222 Murine Bone Marrow Pro-B Cells (along with CALLA/CD10/neutral endopeptidase) of the zinc-dependent metallopeptidase family expressed on early B lineage cells (20) .\n","0.8086322546005249:\tAcquisition of 30F1 expression is indicative of cells at the stage of DJ rearrangement as distinguished from 30F1 -cells completely lacking rearrangement . Subsequently, such cells further differentiate to express BM and this coincides with alteration of the physiological characteristics of the B-lineage cells.\n","\n","Citation context: At the same time, PP strategies are complex and risky. They are technologically complex because they require changes in systems, processes, and products (Aragon-Correa and Sharma, 2003) ; socially complex because they involve diverse stakeholders at different levels (Russo and Fouts, 1997) ; and structurally complex because they require managerial commitment and cross-functional coordination (Aragon-Correa, 1998 ).\n","0.7198380827903748:\tComplexity creates suboptimization, poor imitability, and superior rents for firms capable of generating the appropriate resources and capabilities (Amit & Schoemaker, 1993) . Proactive environmental strategies are socially complex because they require integration of a number of external and internal views (Marcus & Gelfen, 1998; Russo & Fouts, 1997) .\n","0.7041639685630798:\tProactive environmental strategies are socially complex because they require integration of a number of external and internal views (Marcus & Gelfen, 1998; Russo & Fouts, 1997) . These views are often contradictory and/or difficult to understand.\n","0.701062798500061:\tFor instance, product stewardship requires integrating the external and internal perspectives of stakeholders into product design and development processes (Hart, 1995) . These radical changes are more difficult to make in a complex business environment, in which organizations simply tend to make small adjustments (Smart & Vertinsky, 1984) .\n","0.6961365342140198:\tHowever, these resources and capabilities are also difficult to create, administer, and use/ implement (Black & Boal, 1994) . Best practices and processes associated with proactive environmental strategies require far-reaching organizational and managerial changes, are more systemic than most social issues, and affect a broader constellation of organizational functions (Russo & Fouts, 1997) .\n","\n","Citation context: On the other hand, WPCN studies have recently incorporated full duplex technology allowing the access point and the users achieve simultaneous energy transfer and data communication due to recent advances in self-interference cancellation techniques and their practical implementations under the development of 5G and beyond networks. The authors in - have considered the full duplex models for MLS and STM.\n","0.7724465131759644:\tIn this paper, we have proposed a new protocol to enable simultaneous downlink wireless power transfer (WPT) and uplink information transmission for a wireless communication network with a full-duplex hybrid access point (HAP) and a set of wireless users with energy harvesting capabilities. Time division multiple access (TDMA) is employed to realize the multiuser uplink transmission.\n","0.7510252594947815:\tThe contribution and main results of this paper are listed as follows: • We propose a new model to enable simultaneous downlink WPT and uplink information transmission for a multi-user wireless network by employing a full-duplex HAP.\n","0.7382001876831055:\tIn , the authors studied the tradeoff between information rate and power transfer in a frequency selective wireless system. In , the authors proposed using multiple antennas to achieve simultaneous wireless information and power transfer for the emerging self-powered wireless networks.\n","0.7074273824691772:\tIn , the authors proposed using multiple antennas to achieve simultaneous wireless information and power transfer for the emerging self-powered wireless networks. In , the optimal power splitting between information decoding and energy harvesting was derived to minimize the outage probability.\n","\n","Citation context: Based on fragment screening, JF-31-MG46, compounds 27 and 29, and compound 8 of 2,1-benzothiazine 2,2-dioxides were identified to bind in the \"N pocket\". These compounds were proposed to hinder RdRp conformational changes during its transition from initiation to elongation, and thus inhibit RdRp activity .\n","0.8710998296737671:\tGenome-sequencing of compound-resistant DENV replicons, identified amino acid changes that mapped to the N pocket. Since inhibitors bind at the thumb/palm interface of the RdRp, this class of compounds is proposed to hinder RdRp conformational changes during its transition from initiation to elongation.\n","0.8387553691864014:\tSince inhibitors bind at the thumb/palm interface of the RdRp, this class of compounds is proposed to hinder RdRp conformational changes during its transition from initiation to elongation. This is the first report of a class of pan-serotype and cell-active DENV RdRp inhibitors.\n","0.8362516164779663:\tPresumably, retraction of the priming loop (aa782-809) from the active site during enzyme elongation alters the conformation of the N-pocket, leading to weaker binding affinities of the RdRp for the compounds. DENV RdRp N-pocket compounds discovered here, share some common features with Site III non-nucleoside inhibitors described earlier for the HCV polymerase .\n","0.8037520051002502:\tFollowing fragment-based screening using X-ray crystallography, we identified 3, a bi-phenyl acetic acid fragment, that bound to a pocket in the DENV3 RdRp thumb subdomain (IC 50 734 μM; Fig 1 and Table 1 ; 37). Iterative rounds of structure-guided design led to compounds that inhibited both DENV polymerase activity and viral replication in cells (Fig 1 and Table 1 ; Fig 1A and 1B in S1 Text) .\n","\n","Citation context: Recently, it was proposed that RNA binding showed no preference between NS5 and RdRp, since both of these proteins bound RNA through the same interacting sequence. Furthermore, the MTase exhibited negligible RNA specificity .\n","0.8783941268920898:\tNo preference for RNA binding was observed between NS5 versus RdRp because in every case where the same interacting sequence was seen at least twice, both NS5 and RdRp were identified as interaction partners. In other words, the methylase domain likely had negligible RNA specificity.\n","0.8177430629730225:\tBecause negative strand flaviviral RNA is thought to exist primarily in duplex form in infected cells, RdRp interactions with this strand might not predominate. No preference for RNA binding was observed between NS5 versus RdRp because in every case where the same interacting sequence was seen at least twice, both NS5 and RdRp were identified as interaction partners.\n","0.763001561164856:\tIn other words, the methylase domain likely had negligible RNA specificity. Finally, the genomic interaction sequences revealed significant correlation (p Ͻ 5E-11, see \"Experimental Procedures\") with recombination regions documented in DENV-1 (14) and JEV (15) studies and also contained sites at which 5Ј-and 3Ј-DENV regions joined together to form defective interfering (DI) particles (16) .\n","0.7289167642593384:\tNumerous RNA variants, including some residing on the 5Ј-SLA, are not included on the table, as their interactions with RdRp were particularly weak. B, analysis of 3Ј-SL-TL's tolerance for mutations in Arg-773.\n"]}]},{"cell_type":"markdown","source":["Some outputs have no citation contexts"],"metadata":{"id":"Io3zYngLbSSg"}},{"cell_type":"code","source":["### TODO: Find distortion scores for all citations\n","cc = citation_contexts[0]\n","### (1) Find citation's context_window:\n","# Find the citing paper key (secondary)\n","citing_id = cc['citing_paper_id']\n","print(\"citing: \" + str(citing_id))\n","# Extract citation context (context size is hyperparam)\n","if context_type_paragraph: # If context window is paragraph\n","    context_window = cc['context']\n","else: # Context paragraph is something else\n","    # Collect PDF text of the citing paper\n","    pdf_parse = paper_id_to_pdf_parse[citing_id]\n","    paragraphs = pdf_parse['abstract'] + pdf_parse['body_text']\n","    citing_sentences = []\n","    # Loop through the entire text\n","    context_sent_index = 0\n","    for paragraph in paragraphs: # This can (and maybe should...) be done in preprocessing block\n","        doc = nlp(paragraph['text'])\n","        if paragraph['text'] == cc['context']:\n","            citation_start_index = cc['citation_mention_start']\n","            for sent in doc.sents:\n","                if citation_start_index > sent.start_char and citation_start_index < sent.end_char:\n","                    context_sent_index = len(citing_sentences)\n","                citing_sentences.append(sent.text)\n","        else:\n","            for sent in doc.sents:\n","                citing_sentences.append(sent.text)\n","    try:\n","        context_window = ' '.join(citing_sentences[context_sent_index + context_offset : context_sent_index + context_offset + context_size]) # this may causes errors with offset? start of array. also end\n","    except:\n","        context_window = ' '.join(citing_sentences[context_sent_index : context_sent_index + context_size])\n","\n","### (2) Generate array of all sentences in the cited paper\n","cited_id = cc['cited_paper_id']\n","print(\"cited: \" + str(cited_id))\n","pdf_parse = paper_id_to_pdf_parse[cited_id]\n","paragraphs = pdf_parse['abstract'] + pdf_parse['body_text']\n","sentences = []\n","for paragraph in paragraphs:\n","    doc = nlp(paragraph['text'])\n","    for sent in doc.sents:\n","        sentences.append(sent.text)\n","\n","# print(\"\\n\\tCiting / Secondary Paper:\")\n","# for s in citing_sentences:\n","#   print(s)\n","\n","# print(\"\\n\\tCited / Primary Paper:\")\n","# for s in sentences:\n","#   print(s)\n","\n","### (3) Score all windows in cited/primary\n","cited_windows = []\n","cited_scores = []\n","# For each possible window in cited/primary:\n","for i in range(len(sentences) - window_size + 1):\n","    # Extract window from cited paper\n","    curr_window = ' '.join(sentences[i:i + window_size])\n","    # Conduct similarity assessment\n","    score = similarity(scoring_type, context_window, curr_window)\n","    cited_windows.append(curr_window)\n","    cited_scores.append(score)\n","\n","### (4) Determine windows in top-k/threshold\n","cited_scores, cited_windows = sort_windows(cited_scores, cited_windows, threshold, k)\n","print(\"\\nCitation context: \" + context_window)\n","for i in range(len(cited_scores)):\n","    print(str(cited_scores[i]) + \":\\t\" + cited_windows[i])\n","\n","### (5) TODO: Find distortion scores for each window\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"outputId":"cf93cb64-3dd0-4586-c0cf-635357845ee5","id":"KJ4TCUVvbokI","executionInfo":{"status":"ok","timestamp":1675146023176,"user_tz":480,"elapsed":3385,"user":{"displayName":"Maxwell Bennett","userId":"04059206930111574135"}}},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["citing: 18980380\n","cited: 18980380\n","\n","Citation context: Notice that a classical technique called \"lifting\" is used here: We introduce an auxiliary random vectorũ, so that some non-linear relationship can be modeled linearly. For example, a constraint on the variance can be modeled using this standard form (see [22, Example 2] ), which is otherwise impossible without the auxiliary variable.\n","1.0:\tNotice that a classical technique called \"lifting\" is used here: We introduce an auxiliary random vectorũ, so that some non-linear relationship can be modeled linearly. For example, a constraint on the variance can be modeled using this standard form (see [22, Example 2] ), which is otherwise impossible without the auxiliary variable.\n","0.5523049645390071:\tThe set of joint distribution of (p s , r s ) is hence C s Δ = (ps ,rs)C s . Notice that a classical technique called \"lifting\" is used here: We introduce an auxiliary random vectorũ, so that some non-linear relationship can be modeled linearly.\n","0.5274760485757314:\tFor example, a constraint on the variance can be modeled using this standard form (see [22, Example 2] ), which is otherwise impossible without the auxiliary variable. This lifting technique thus allows us to model a rich variety of structural information about the marginal distribution of (p, r) in a unified manner.\n","0.165272856130404:\tThe convexity can be easily shown, which is omitted due to space constraints (see [22] for details). To show the compactness, notice thatC s is weakly closed (i.e., closed w.r.t.\n"]}]},{"cell_type":"code","source":["different = 0\n","same = 0\n","for i in range(len(citation_contexts)):\n","  if citation_contexts[i]['citing_paper_id'] != citation_contexts[i]['cited_paper_id']:\n","    different += 1\n","  else:\n","    same += 1\n","print(\"Different: \" + str(different))\n","print(\"Same: \" + str(same))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Rbn8lHfPNwnj","executionInfo":{"status":"ok","timestamp":1675146287349,"user_tz":480,"elapsed":274,"user":{"displayName":"Maxwell Bennett","userId":"04059206930111574135"}},"outputId":"48d64351-8478-47c2-ce04-a6dd59c46e27"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Different: 10106\n","Same: 702\n"]}]},{"cell_type":"code","source":[],"metadata":{"id":"502_W_rlfKkE"},"execution_count":null,"outputs":[]}]}